internal/notifier/forwarder_test.go

```diff
@@ -25,42 +25,113 @@ import (
 	"net/http"
 	"net/http/httptest"
 	"testing"
+	"time"
 
 	fuzz "github.com/AdaLogics/go-fuzz-headers"
 	"github.com/fluxcd/pkg/runtime/events"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	"github.com/stretchr/testify/require"
 )
 
+func TestForwarder_New(t *testing.T) {
+	tests := []struct {
+		name    string
+		hmacKey []byte
+		err     bool
+	}{
+		{
+			name:    "nil HMAC key passes",
+			hmacKey: nil,
+			err:     false,
+		},
+		{
+			name:    "empty HMAC key fails",
+			hmacKey: []byte{},
+			err:     true,
+		},
+		{
+			name:    "happy path with HMAC key from empty string",
+			hmacKey: []byte(""),
+			err:     true,
+		},
+		{
+			name:    "non-empty HMAC key adds signature header",
+			hmacKey: []byte("7152fed34dd6149a7c75a276c510da27cb6f82b0"),
+			err:     false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := NewForwarder("http://example.org", "", nil, nil, tt.hmacKey)
+			if tt.err {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
 func TestForwarder_Post(t *testing.T) {
-	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		b, err := io.ReadAll(r.Body)
-		require.NoError(t, err)
-
-		require.Equal(t, "source-controller", r.Header.Get("gotk-component"))
-		require.Equal(t, "token", r.Header.Get("Authorization"))
-		var payload = events.Event{}
-		err = json.Unmarshal(b, &payload)
-		require.NoError(t, err)
-		require.Equal(t, "webapp", payload.InvolvedObject.Name)
-		require.Equal(t, "metadata", payload.Metadata["test"])
-	}))
-	defer ts.Close()
-
-	headers := make(map[string]string)
-	headers["Authorization"] = "token"
-	forwarder, err := NewForwarder(ts.URL, "", headers, nil)
-	require.NoError(t, err)
-
-	err = forwarder.Post(context.TODO(), testEvent())
-	require.NoError(t, err)
+	tests := []struct {
+		name       string
+		hmacKey    []byte
+		hmacHeader string
+	}{
+		{
+			name:       "happy path with nil HMAC key",
+			hmacKey:    nil,
+			hmacHeader: "",
+		},
+		{
+			name:       "non-empty HMAC key adds signature header",
+			hmacKey:    []byte("7152fed34dd6149a7c75a276c510da27cb6f82b0"),
+			hmacHeader: "sha256=65b018549b1254e7226d1c08f9567ee45bc9de0fc4e7b1a40253f9a018b08be7",
+		},
```

Comment:
We could add a test to ensure that the headers input does not overwrite the signature created by the new code.

Reply:
Great idea. I added some more testing around the `X-Signature` header.